src/Limiters/limiter.jl

```diff
@@ -0,0 +1,125 @@
+        ρ_e_mass = sum(ρ_e_slab)
+        ρq_e_mass = sum(ρq_e_slab)
+
+        # This should never happen, but if it does, don't limit
+        if ρ_e_mass <= zero(FT)
+            continue
+        end
+
+        # Relax constraints to ensure limiter has a solution:
+        # This is only needed if running with the SSP CFL>1 or
+        # due to roundoff errors.
+        if ρq_e_mass < min_ρq[e] * ρ_e_mass
+            min_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        if ρq_e_mass > max_ρq[e] * ρ_e_mass
+            max_ρq[e] = ρq_e_mass / ρ_e_mass
+        end
+
+        local_geometry_slab = Fields.slab(space.local_geometry, e)
+
+        q_data .= ρq_e_data ./ ρ_e_data
+
+        # Weighted least squares problem iteration loop
+        for iter in 1:(Nq * Nq - 1)
+            mass_change = zero(FT)
+            # Iterate over quadrature points
+            for j in 1:Nq, i in 1:Nq
+                ρ_wJ_data[i, j] = local_geometry_slab[i, j].WJ .* ρ_e_data[i, j]
+                # Compute the error tolerance and project q into the
+                # upper and lower bounds
+                if q_data[i, j] > max_ρq[e]
+                    mass_change += (q_data[i, j] - max_ρq[e]) * ρ_wJ_data[i, j]
+                    q_data[i, j] = max_ρq[e]
+                elseif q_data[i, j] < min_ρq[e]
+                    mass_change -= (min_ρq[e] - q_data[i, j]) * ρ_wJ_data[i, j]
+                    q_data[i, j] = min_ρq[e]
+                end
+            end
+
+            # By this projection of q to the upper/lower bounds, we
+            # have introduced a mass_change. Check if we are within the chosen
+            # tolerance
+            if abs(mass_change) <= rtol * abs(ρq_e_mass)
+                break
+            end
+
+            weights_sum = zero(FT)
+
+            # If the change was positive, the removed mass is added
+            if mass_change > 0
+                # Iterate over quadrature points
+                for j in 1:Nq, i in 1:Nq
+                    if q_data[i, j] < max_ρq[e]
+                        weights_sum += ρ_wJ_data[i, j]
+                    end
+                end
+                for j in 1:Nq, i in 1:Nq
+                    if q_data[i, j] < max_ρq[e]
+                        q_data[i, j] += mass_change / weights_sum
+                    end
+                end
+            else # If the change was negative, the added mass is removed
+                # Iterate over quadrature points
+                for j in 1:Nq, i in 1:Nq
+                    if q_data[i, j] > min_ρq[e]
+                        weights_sum += ρ_wJ_data[i, j]
+                    end
+                end
+                for j in 1:Nq, i in 1:Nq
+                    if q_data[i, j] > min_ρq[e]
+                        q_data[i, j] += mass_change / weights_sum
+                    end
+                end
+            end
+        end
+        ρq_e_data .= q_data .* ρ_e_data
+    end
+end
```

Comment:
It's generally good to return the thing you've modified:
```suggestion
    return ρq
end
```